### DISCUSSION TITLE
Some questions about YesSQL in OrchardCore and how do we rollback if there is exception?

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#7754] [Q&A]
Hello everyone, 
I have 2 questions related to YesSql in OrchardCore.

**Here is my 1st question:**
Am I understand these correctly?
- YesSql.ISession is registered as per web request (AddScoped) in https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore/OrchardCore.Data/OrchardCoreBuilderExtensions.cs#L102
- Given this scenario: 
   - We inject ISession to a Controller and save multiple items in an action method
   - When a request end, Microsoft.Extensions.DependencyInjection will dispose the ISession object.
   - ISession object and inside dispose of ISession method will call commit transaction or rollback if there is session.Cancel call 
   - Refer to https://github.com/sebastienros/yessql/blob/5731eede78d879fb49f458ebd61df1fdafb0010a/src/YesSql.Core/Session.cs#L654

**2nd question**
Does current implementation of OrchardCore auto rollback if there is an exception in an action method or do we have any action filter to cancel a session if there is an exception? 
I think we need to do something like this.

```
try{
    session.Save(xxx);
    session.Save(yyy);
}catch(Exception ex){
    session.Cancel();
}
```

Thanks.

**Comment by jtkech** [comment]
@aaronamm 

> YesSql.ISession is registered as per web request

More precisely per `ShellScope`, e.g. we can create child scopes or a scope on another tenant this in the same http request scope.

> ISession object and inside dispose of ISession method will call commit transaction or rollback if there is session.Cancel

This is before disposing the `ShellScope` that we call asynchronously the session `CommitAsync()` so before the session is disposed.

So, all is done by `using` and in the context of a `ShellScope`, if there is an exception currently what we ensure is that the session is disposed (so here committed synchronously by its dispose), but you're right the session is not canceled.

Hmm, easy to add but in some cases we may want to not cancel the session, assuming that an entity need to be validated before doing a session save on it, needs more thinkings.

But you're right, if you need it in your scenario, currently you would need to use your own try catch blocks

**Reply by aaronamm** [reply]
@jtkech Thank you for prompt reply and some useful information.
